designs/0005-static-matrices.md

```diff
@@ -0,0 +1,240 @@
+regardless of how the language was handled.
+
+At this point, the Math maturing to the point that it is possible to compile a model using
+varmats values with a development version of the Stanc3 compiler (available
+[here](https://github.com/stan-dev/stanc3/pull/755), though it may take some effort to run since
+it is a work in progress). This Stanc3 compiler defines any variable with a name ending in
+`_varmat` to use a varmat type.
+
+A basic bernoulli MRP model with five population level covariates and seven group terms is available
+[here](https://github.com/bbbales2/computations_in_glms/blob/master/benchmark_model/mrp_varmat.stan).
+
+The initial varmat implementation of this model achieves a per-gradient speedup of around two.
+
+Now that varmat variables can be assigned, it is possible that the compiler silently replace every
+matvar in Stan with a varmat, and none of the standard Stan use cases would be affected.
+
+There are four major implementation strategies available at the Stanc3 level.
+
+1. Implement a new Stan type to differentiate explicitly between matvar and varmat types
+
+2. Use matvar types whenever there is any function in the Stan program that does not support varmat
+
+3. Use varmat types by default, but if variable assignment is detected, or a varmat is used
+in an unsupported function convertion to and from matvar is used automatically.
+
+4. Use varmat types everywhere
+
+The original proposal was leaning towards #1 because it most naturally corresponds to the Math
+implementation. However, this leaves the problem of explicitly declaring varmat to the user. To
+avoid duplicating the user's manual and function reference for varmat and matvar functions, there
+would also need to be some automatic, under the hood conversions between the types. This automatic
+converting would make it so that if a varmat is forgotten anywhere, the entire program will be
+slowed down. This sort of explicit typing shouldn't be too much of a problem in development models
+or generated Stan code, but in large, user-written models this would be error prone to implement
+and maintain.
+
+The advantage of #2 is that it would allow programs that can be written using only varmat functions
+to be very fast (and not even worry about any conversion to matvar or not). The problem with this
+implementation is that, the difference between `varmat` and `matvar` would again need documented
+at the user level and some notation would need added to the user manual to indicate if a function
+is or is not available for optimization.
+
+The advantage of #3 is that varmat and matvar types could be implemented without any extra
+documentation for the user. Similarly to #2, if the user uses functions only available for matvar,
+there will be some slowdown as conversions between varmat and matvar are handled, but unlike #2
+the slowdown might not be catastrophic. In this case it would not be as crucial to carefully
```

Comment:
Which of 2/3 is slower is just speculation. Also I bet it will depend on the model a lot. I would prefer this removed unless you support it by benchmarking multiple models.

Reply:
Hmm, it is, but also I don't want to say 2 and 3 are the same, cause they're not. The reason to do 3 would be to avoid at least one of:

1. catastrophic slowdowns in 2 where using a weird function with no varmat implementation in one place means the whole program is slow

2. documenting all functions that do and do not support varmat and to what extent (only certain arguments, etc.)

Reply:
In the end I added the assumption that the matvar parts are cheap. So like, if someone needs one `lbeta` call and that isn't a `varmat` function, then all the matrix operations and stuff don't get converted back to `matvar` -- only the single, assumed relatively cheap function gets wrapped up.